Enhancement of SSIM/MS-SSIM and BRISQUE, Refactoring of tests for SSIM/MS-SSIM and BRISQUE

denproc opened the pull request:
Closes #126 
Closes #118

## Proposed Changes
- Compare SSIM/MS-SSIM with TensorFlow implementation using real images;
- Compare BRISQUE with another `pybrisque` implementation using real images;
- Refactoring of SSIM/MS-SSIM and BRISQUE;
- Refactoring of tests for SSIM/MS-SSIM and BRISQUE;
- Fixed memory consumption during SSIM/MS-SSIM testing;
- Added better testing on GPU for SSIM/MS-SSIM and BRISQUE.

denproc commented:
@snk4tr @zakajd ready for review

snk4tr commented on piq/ssim.py:
```diff
@@ -144,66 +142,40 @@ def __init__(self, kernel_size: int = 11, kernel_sigma: float = 1.5, k1: float =
-        return {'mean': loss.mean,
-                'sum': loss.sum
-                }[self.reduction](dim=0)
+        score = ssim(x=prediction, y=target, kernel_size=self.kernel_size, kernel_sigma=self.kernel_sigma,
+                     data_range=self.data_range, reduction=self.reduction, full=False, k1=self.k1, k2=self.k2)
+        return torch.ones_like(score) - score
```
Very neat! Like it 👍

snk4tr commented on piq/ssim.py:
```diff
@@ -433,48 +362,44 @@ def _ssim_per_channel(x: torch.Tensor, y: torch.Tensor, kernel: torch.Tensor, da
     # Set alpha = beta = gamma = 1.
     cs_map = (2 * sigma12 + c2) / (sigma1_sq + sigma2_sq + c2)
-    ssim_map = ((2 * mu1_mu2 + c1) / (mu1_sq + mu2_sq + c1)) * cs_map
+    ssim_map = (2 * mu1_mu2 + c1) / (mu1_sq + mu2_sq + c1) * cs_map
```
I understand that removing these parentecies are redundant since they do not change the order of operations. However, I would still leave them to make this order more explicit.

denproc replied:
I suggest to keep it as is. Pytorch 1.6.0 will support complex values, which will lead to possible refactoring of the logic for complex valued methods.

snk4tr commented on piq/ssim.py:
```diff
@@ -487,30 +412,44 @@ def _ssim(x: torch.Tensor, y: torch.Tensor, kernel: torch.Tensor, data_range: Un
+    for iteration in range(levels):
+        if iteration > 0:
+            padding = (x.shape[2] % 2, x.shape[3] % 2)
+            x = f.pad(x, pad=[padding[0], 0, padding[1], 0], mode='replicate')
+            y = f.pad(y, pad=[padding[0], 0, padding[1], 0], mode='replicate')
+            x = f.avg_pool2d(x, kernel_size=2, padding=0)
+            y = f.avg_pool2d(y, kernel_size=2, padding=0)
+
```
There is the same code fragment in the `_multi_scale_ssim_complex` function. Consider extracting this duplication into a function (optional)

denproc replied:
I suggest to keep it as is. Pytorch 1.6.0 will support complex values, which will lead to possible refactoring of the logic for complex valued methods.

snk4tr commented on piq/ssim.py:
```diff
@@ -487,30 +412,44 @@ def _ssim(x: torch.Tensor, y: torch.Tensor, kernel: torch.Tensor, data_range: Un
     ssim_val = None
-    for _ in range(levels):
+    for iteration in range(levels):
+        if iteration > 0:
```
Why do you need this condition?

zakajd replied:
To compute SSIM on original scale before downsampling it

denproc replied:
Just changed order of operation to get rid of final downscale, the logic remains the same.

snk4tr commented on piq/ssim.py:
```diff
@@ -625,46 +546,61 @@ def _ssim_complex(x: torch.Tensor, y: torch.Tensor, kernel: torch.Tensor, data_r
         y_real = y[..., 0]
         y_imag = y[..., 1]
-        mcs.append(cs)
+        if iteration > 0:
```
The same question here: why do you need this condition?

denproc replied:
The same answer here :)

snk4tr commented on tests/test_ssim.py:
```diff
@@ -2,96 +2,81 @@
 @pytest.fixture(scope='module')
 def prediction() -> torch.Tensor:
-    return torch.rand(3, 3, 256, 256)
+    return torch.rand(3, 3, 161, 161)
```
Is this a minimal size for MS-SSIM? 
If so, may be add it somewhere in the documentation?

denproc replied:
Yes, we have limitations due to downscaling inside the MS-SSIM. Added to doc.

zakajd commented on piq/brisque.py:
```diff
@@ -11,9 +11,114 @@
 from torch.nn.modules.loss import _Loss
 from torch.utils.model_zoo import load_url
-import torch.nn.functional as F
+import torch.nn.functional as f
```
It's very common to import `functional` module as `F`. Why do you want to change this?

zakajd commented on tests/test_brisque.py:
```diff
@@ -3,96 +3,87 @@
 @pytest.fixture(scope='module')
 def prediction_grey() -> torch.Tensor:
-    return torch.rand(3, 1, 256, 256)
+    return torch.rand(3, 1, 128, 128)
```
Yes, I'm using 96x96 now. It works fine for most cases

zakajd commented on tests/test_brisque.py:
```diff
@@ -3,96 +3,87 @@
 
-def test_brisque_all_zeros_or_ones() -> None:
-    size = (1, 1, 256, 256)
+def test_brisque_all_zeros(device: Any) -> None:
+    size = (1, 1, 128, 128)
```
Test name is misleading, as you also check tensor of ones inside. 
Consider doing this using `pytest.mark.parametrize`

```suggestion
@pytest.mark.parametrize(
    "input,expectation",
    [(torch.zeros(2, 1, 96, 96), AssertionError),
      (torch.ones(2, 1, 96, 96), AssertionError)],
)
def test_brisque_for_special_cases(input: torch.Tensor, expectation: Any, device: str) -> None:
    with expectation:
        brisque(input.to(device), reduction='mean')

```

zakajd commented on piq/utils.py:
```diff
@@ -18,7 +18,7 @@ def _adjust_dimensions(input_tensors: Union[torch.Tensor, Tuple[torch.Tensor, to
         if tmp.dim() != 4 and tmp.dim() != 5:
             raise ValueError(f'Expected 2, 3, 4 or 5 dimensions (got {tensor.dim()})')
-        resized_tensors.append(tmp)
+        resized_tensors.append(tmp.type(torch.float32))
```
Why we need to change tensor type during resize? 
This should be mentioned in docstring

snk4tr replied:
Yeah, I noticed this one as well. 
We have not addressed data types of passed tensors and to me it seems like too much of work to do that now (because then all data types will need to be tested etc.)

I agree with @zakajd that if you change data type it need to be explicitly defined several times in different places or (even better) just be avoided. What is the use case when the absence of this conversion break things down?

denproc replied:
This is the case for SSIM and MS-SSIM. We usually use `.to(prediction)`, but it doesn't work here in case of prediction with `dtype=torch.uint8`. For example, it is impossible to create gaussian filter with `dtype=torch.uint8`. 
Next, we will have:
- the rounded result is equal `1` instead of `0.6824` in case of SSIM;
- conflict at `torch.nn.functional.pad` in case of MS-SSIM.

Almost all operations return `float` results. The translation to `torch.float32` is already used as a result of `x = x / float(data_range)`.
Since we implicitly change the type of the input tensors, I propose to make it at `_validation_input` explicitly and show the corresponding warning.

snk4tr replied:
Agh I would still avoid doing it.   
If one has `torch.ByteTensor` prediction, let him/her convert it to `torch.float32` and then compute SSIM/MS-SSIM with it. This whole type conversion thing is out of the scope of this lib.

denproc replied:
@snk4tr, `torch.ByteTensor` is `torch.uint8`, which is usual data type for images. Let's take `FSIM`. We can pass `torch.uint8` tensor and got the answer. This is the case for all metrics meaning that we already provide user with such implicit conversion. What is more, it is not logical making one metrics unavailable because of wrong type and keeping other metrics work with it. I move the conversion especially for torch.uint8 inside SSIM.

denproc replied:
@snk4tr @zakajd What do you think about strict assert of the input data type? One can pass the `ByteTensor` to a measure, but they cannot use them for back propagation (gradients are available only for `FloatTensor`).

zakajd commented on tests/test_ssim.py:
```diff
@@ -2,96 +2,81 @@
 import tensorflow as tf
-
+import numpy as np
+from PIL import Image
```
Consider using `from skimage.io import imread` instead of `PIL.Image`. It's slightly faster, reads images directly into np.arrays and already included in testing dependencies.

denproc commented:
### SSIM/MS-SSIM
- I just optimised the code, removed unnecessary functions, iterations and so on. 
- Also, I added padding with reflection mode instead of zero-padding inside `avg_pool2d` and `conv2d`, which makes our implementation closed to TF and original MATLAB implementations.

### Tests
- Tests were optimised to reduce the duplications. It eliminated some inconsistencies.
- Added tests, which use real images.

In general, the implementation is ok. The main sources of possible errors are float precision and zero-paddings instead of reflect/symmetric.

denproc commented:
@zakajd @snk4tr ready to merge

snk4tr commented:
@denproc since #131 was merged you need to resolve conflicts first

denproc commented:
@snk4tr done
